unit.go

```diff
@@ -0,0 +1,338 @@
+// Copyright ©2013 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package unit
+
+import "fmt"
+
+// Unit is a type a value with generic SI units. Most useful for
+// translating between dimensions, for example, by multiplying
+// an acceleration with a mass to get a force. Please see the
+// package documentation for further explanation.
+type Unit struct {
+	dimensions Dimensions // Map for custom dimensions
+	formatted  string
+	value      float64
+}
+
+// New creates a new variable of type Unit which has the value
+// specified by value and the dimensions specified by the
+// base units struct. The value is always in SI Units.
+//
+// Example: To create an acceleration of 3 m/s^2, one could do
+// myvar := CreateUnit(3.0, &Dimensions{unit.LengthDim: 1, unit.TimeDim: -2})
+func New(value float64, d Dimensions) *Unit {
+	u := &Unit{
+		dimensions: make(map[Dimension]int),
+		value:      value,
+	}
+	for key, val := range d {
+		if val != 0 {
+			u.dimensions[key] = val
+		}
+	}
+	return u
+}
+
+// Clone makes a copy of a unit.
+func (u *Unit) Clone() *Unit {
+	o := &Unit{
+		dimensions: make(map[Dimension]int),
+		value:      u.value,
+	}
+	for key, val := range u.dimensions {
+		if val != 0 {
+			o.dimensions[key] = val
+		}
+	}
+	return o
+}
+
+// DimensionsMatch checks if the dimensions of two *Units are the same.
+func DimensionsMatch(a, b *Unit) bool {
+	aUnit := a
+	bUnit := b
+	if len(aUnit.dimensions) != len(bUnit.dimensions) {
+		return false
+	}
+	for key, val := range aUnit.dimensions {
+		if bUnit.dimensions[key] != val {
+			return false
+		}
+	}
+	return true
+}
+
+// Matches checks if the two sets of dimensions are the same.
+func (d Dimensions) Matches(d2 Dimensions) bool {
+	if len(d) != len(d2) {
+		return false
+	}
+	for key, val := range d {
+		if d2[key] != val {
```

Comment:
what if `d2` does not have key `key` but another one that `d` hasn't ?
shouldn't we also use `v2,ok := d2[key]` and/or also range-loop over `d2` ?